Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+State.swift

```diff
@@ -0,0 +1,242 @@
+            switch self.state {
+            case .http1(var http1StateMachine):
+                return self.state.modify { state -> Action in
+                    let action = http1StateMachine.failedToCreateNewConnection(error, connectionID: connectionID)
+                    state = .http1(http1StateMachine)
+                    return state.modify(with: action)
+                }
+
+            case .modify:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        mutating func waiterTimeout(_ requestID: RequestID) -> Action {
+            switch self.state {
+            case .http1(var http1StateMachine):
+                return self.state.modify { state -> Action in
+                    let action = http1StateMachine.timeoutWaiter(requestID)
+                    state = .http1(http1StateMachine)
+                    return state.modify(with: action)
+                }
+
+            case .modify:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        mutating func cancelWaiter(_ requestID: RequestID) -> Action {
+            switch self.state {
+            case .http1(var http1StateMachine):
+                return self.state.modify { state -> Action in
+                    let action = http1StateMachine.cancelWaiter(requestID)
+                    state = .http1(http1StateMachine)
+                    return state.modify(with: action)
+                }
+
+            case .modify:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        mutating func connectionTimeout(_ connectionID: Connection.ID) -> Action {
+            switch self.state {
+            case .http1(var http1StateMachine):
+                return self.state.modify { state -> Action in
+                    let action = http1StateMachine.connectionTimeout(connectionID)
+                    state = .http1(http1StateMachine)
+                    return state.modify(with: action)
+                }
+
+            case .modify:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        /// A connection has been closed
+        mutating func connectionClosed(_ connectionID: Connection.ID) -> Action {
+            switch self.state {
+            case .http1(var http1StateMachine):
+                return self.state.modify { state -> Action in
+                    let action = http1StateMachine.connectionClosed(connectionID)
+                    state = .http1(http1StateMachine)
+                    return state.modify(with: action)
+                }
+
+            case .modify:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        mutating func http1ConnectionReleased(_ connectionID: Connection.ID) -> Action {
+            guard case .http1(var http1StateMachine) = self.state else {
+                preconditionFailure("Invalid state")
```

Comment:
I would generally recommend throwing in cases where we believe it's possible to trigger the event accidentally: i.e. because someone using the program got confused, or from the network. In this case I believe the precondition is here because an outer state machine should prevent us getting here at all.

Nonetheless, I would recommend we switch exhaustively over the states, even if we crash in the other cases.